File: EngineSource/Objects/Components/PhysicsComponent.cpp
#include "PhysicsComponent.h"
#include <Math/Physics/Physics.h>
#include <iostream>

void PhysicsComponent::Begin()
{
}

void PhysicsComponent::Destroy()
{
	Physics::PhysicsBody* Body = static_cast<Physics::PhysicsBody*>(PhysicsBodyPtr);
	if (Body)
	{
		SetActive(true);
		Physics::RemoveBody(Body, true);
		delete Body;
	}
}

void PhysicsComponent::Update()
{
	if (!PhysicsBodyPtr)
	{
		return;
	}

	if (!GetParent())
	{
		return;
	}

	Physics::PhysicsBody* Body = static_cast<Physics::PhysicsBody*>(PhysicsBodyPtr);

	if (Body->ColliderMovability == Physics::MotionType::Static && Active)
	{
		Transform ComponentTransform = Component::GetWorldTransform();
		Body->SetPosition(ComponentTransform.Position);
		Body->SetRotation(ComponentTransform.Rotation);
	}
}


void PhysicsComponent::CreateBox(Transform RelativeTransform, Physics::MotionType BoxMovability, Physics::Layer CollisionLayers)
{
	if (PhysicsBodyPtr)
	{
		Destroy();
	}

	this->RelativeTransform = RelativeTransform;

	Transform ComponentTransform = Component::GetWorldTransform();

	Physics::PhysicsBody* Body = new Physics::BoxBody(ComponentTransform.Position,
		ComponentTransform.Rotation,
		ComponentTransform.Scale,
		(Physics::MotionType)BoxMovability,
		CollisionLayers,
		this);

	
	PhysicsBodyPtr = Body;
	SetActive(true);
}

void PhysicsComponent::CreateSphere(Transform RelativeTransform, Physics::MotionType BoxMovability, Physics::Layer CollisionLayers)
{
	if (PhysicsBodyPtr)
	{
		Destroy();
	}

	this->RelativeTransform = RelativeTransform;

	Transform ComponentTransform = Component::GetWorldTransform();
	
	Physics::PhysicsBody* Body = new Physics::SphereBody(ComponentTransform.Position,
		ComponentTransform.Rotation,
		(ComponentTransform.Scale.X + ComponentTransform.Scale.Y + ComponentTransform.Scale.Z) / 3,
		BoxMovability,
		CollisionLayers,
		this);


	PhysicsBodyPtr = Body;
	SetActive(true);
}

void PhysicsComponent::CreateCapsule(Transform RelativeTransform, Physics::MotionType CapsuleMovability, Physics::Layer CollisionLayers)
{
	if (PhysicsBodyPtr)
	{
		Destroy();
	}

	this->RelativeTransform = RelativeTransform;

	Transform ComponentTransform = GetWorldTransform();

	Physics::PhysicsBody* Body = new Physics::CapsuleBody(ComponentTransform.Position,
		ComponentTransform.Rotation,
		Vector2(ComponentTransform.Scale.X, ComponentTransform.Scale.Y),
		CapsuleMovability,
		CollisionLayers,
		this);


	PhysicsBodyPtr = Body;
	SetActive(true);
}

Transform PhysicsComponent::GetBodyWorldTransform() const
{
	if (!PhysicsBodyPtr)
	{
		return Transform();
	}
	
	Physics::PhysicsBody* Body = static_cast<Physics::PhysicsBody*>(PhysicsBodyPtr);
	return Transform(Body->GetPosition(), Body->GetRotation(), 1);
}

void PhysicsComponent::SetPosition(Vector3 NewPosition)
{
	if (!PhysicsBodyPtr)
	{
		return;
	}

	Physics::PhysicsBody* Body = static_cast<Physics::PhysicsBody*>(PhysicsBodyPtr);
	Body->SetPosition(NewPosition);
}

void PhysicsComponent::SetRotation(Vector3 NewRotation)
{
	if (!PhysicsBodyPtr)
	{
		return;
	}

	Physics::PhysicsBody* Body = static_cast<Physics::PhysicsBody*>(PhysicsBodyPtr);
	Body->SetRotation(NewRotation);
}

void PhysicsComponent::SetScale(Vector3 NewScale)
{
	if (!PhysicsBodyPtr)
	{
		return;
	}

	Physics::PhysicsBody* Body = static_cast<Physics::PhysicsBody*>(PhysicsBodyPtr);

	Vector3 ScaleDifference = NewScale / Body->BodyTransform.Scale;
	if (ScaleDifference != Vector3(1))
	{
		Body->Scale(ScaleDifference);
	}

}

Vector3 PhysicsComponent::GetVelocity() const
{
	if (!PhysicsBodyPtr)
	{
		return 0;
	}
	Physics::PhysicsBody* Body = static_cast<Physics::PhysicsBody*>(PhysicsBodyPtr);
	return Body->GetVelocity();
}

Vector3 PhysicsComponent::GetAngularVelocity() const
{
	if (!PhysicsBodyPtr)
	{
		return 0;
	}
	Physics::PhysicsBody* Body = static_cast<Physics::PhysicsBody*>(PhysicsBodyPtr);
	return Body->GetAngularVelocity();
}

void PhysicsComponent::SetVelocity(Vector3 NewVelocity)
{
	if (!PhysicsBodyPtr)
	{
		return;
	}
	Physics::PhysicsBody* Body = static_cast<Physics::PhysicsBody*>(PhysicsBodyPtr);
	Body->SetVelocity(NewVelocity);
}

void PhysicsComponent::SetAngularVelocity(Vector3 NewVelocity)
{
	if (!PhysicsBodyPtr)
	{
		return;
	}
	Physics::PhysicsBody* Body = static_cast<Physics::PhysicsBody*>(PhysicsBodyPtr);
	Body->SetAngularVelocity(NewVelocity);
}

void PhysicsComponent::SetActive(bool NewActive)
{
	Physics::PhysicsBody* Body = static_cast<Physics::PhysicsBody*>(PhysicsBodyPtr);
	if (NewActive == Active || !Body)
	{
		return;
	}
	if (NewActive)
	{
		Physics::AddBody(Body);
	}
	else
	{
		Physics::RemoveBody(Body, false);
	}
	Active = NewActive;
}

Physics::HitResult PhysicsComponent::ShapeCast(Transform Start, Vector3 End, Physics::Layer Layers, std::set<WorldObject*> ObjectsToIgnore)
{
	if (!PhysicsBodyPtr)
	{
		return Physics::HitResult();
	}
	Physics::PhysicsBody* Body = static_cast<Physics::PhysicsBody*>(PhysicsBodyPtr);
	return Physics::HitResult::GetAverageHit(Body->ShapeCast(Start, End, Layers, ObjectsToIgnore));

}

Physics::HitResult PhysicsComponent::CollisionCheck(Transform Start, Physics::Layer Layers, std::set<WorldObject*> ObjectsToIgnore)
{
	if (!PhysicsBodyPtr)
	{
		return Physics::HitResult();
	}
	Physics::PhysicsBody* Body = static_cast<Physics::PhysicsBody*>(PhysicsBodyPtr);
	Transform InitialTransform = Body->BodyTransform;
	Body->BodyTransform = Start;
	auto Hit =  Physics::HitResult::GetAverageHit(Body->CollisionTest(Layers, ObjectsToIgnore));

	std::swap(InitialTransform, Body->BodyTransform);

	return Hit;
}

File: CSharp/Engine/Components/PhysicsComponent.cs
namespace Engine;
using Engine.Native;
using System;
using System.Runtime.InteropServices;

/**
* @brief
* A component that simulates physics.
* 
* @ingroup CSharp-Components
*/
public class PhysicsComponent : ObjectComponent
{
	/// The type a physics body can have.
	enum BodyType
	{
		Box,
		Sphere,
		Capsule,
		Mesh,
	};

	public override void Tick()
	{
	}


	private delegate IntPtr NewPhysicsComponentDelegate(
		IntPtr Parent,
		Transform t,
		BodyType Type,
		Collision.PhysicsMotionType Movability,
		Collision.Layer Layers);

	private delegate Transform GetTransformDelegate(IntPtr NativePtr);
	private delegate void SetVector3Delegate(IntPtr NativePtr, Vector3 Vec);
	private delegate Vector3 GetVector3Delegate(IntPtr NativePtr);
	private delegate void SetBoolDelegate(IntPtr NativePtr, [MarshalAs(UnmanagedType.U1)] bool NewActive);
	private delegate Collision.HitResponse ShapeCastDelegate(
		IntPtr NativePtr,
		Transform Start,
		Vector3 End,
		Collision.Layer Layers,
		[MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 1)] IntPtr[] ComponentsToIgnore,
		int Length);

	private delegate Collision.HitResponse CollisionCheckDelegate(
	IntPtr NativePtr,
	Transform Where,
	Collision.Layer Layers,
	[MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 1)] IntPtr[] ComponentsToIgnore,
	int Length);


	private void CreateColliderOfType(BodyType Type,
		Transform RelativeTransform,
		Collision.PhysicsMotionType Movability,
		Collision.Layer CollisionLayers)
	{
		NativePtr = (IntPtr)NativeFunction.CallNativeFunction("NewPhysicsComponent", typeof(NewPhysicsComponentDelegate),
			[
			Parent.NativePtr,
			RelativeTransform,
			Type,
			Movability,
			CollisionLayers
			]);
	}

	/// Sets the position of the physics body, in world space.
	public void SetPosition(Vector3 NewPosition)
	{
		NativeFunction.CallNativeFunction("PhysicsComponentSetPosition", typeof(SetVector3Delegate), [NativePtr, NewPosition]);
	}

	/// Sets the rotation of the physics body, in world space.
	public void SetRotation(Vector3 NewRotation)
	{
		NativeFunction.CallNativeFunction("PhysicsComponentSetRotation", typeof(SetVector3Delegate), [NativePtr, NewRotation]);
	}

	/// Sets the scale of the physics body, in world space.
	public void SetScale(Vector3 NewScale)
	{
		NativeFunction.CallNativeFunction("PhysicsComponentSetScale", typeof(SetVector3Delegate), [NativePtr, NewScale]);
	}

	/// Sets the velocity of the physics body.
	public void SetVelocity(Vector3 NewVelocity)
	{
		NativeFunction.CallNativeFunction("PhysicsComponentSetVelocity", typeof(SetVector3Delegate), [NativePtr, NewVelocity]);
	}

	/// Sets the angular velocity of the physics body.
	public void SetAngularVelocity(Vector3 NewVelocity)
	{
		NativeFunction.CallNativeFunction("PhysicsComponentSetAngularVelocity", typeof(SetVector3Delegate), [NativePtr, NewVelocity]);
	}

	/// Gets the velocity of the physics body.
	public Vector3 GetVelocity()
	{
		return (Vector3)NativeFunction.CallNativeFunction("PhysicsComponentGetVelocity", typeof(GetVector3Delegate), [NativePtr]);
	}

	/// Gets the angular velocity of the physics body.
	public Vector3 GetAngularVelocity()
	{
		return (Vector3)NativeFunction.CallNativeFunction("PhysicsComponentGetAngularVelocity", typeof(GetVector3Delegate), [NativePtr]);
	}

	/// Gets the transform of the simulated physics body, in world space.
	public Transform GetBodyWorldTransform()
	{
		return (Transform)NativeFunction.CallNativeFunction("PhysicsComponentGetTransform", typeof(GetTransformDelegate), [NativePtr]);
	}

	/**
	* @brief
	* Creates a @ref Physics::BoxBody with the given parameters.
	*
	* @param RelativeTransform
	* The Transform of the PhysicsBody, relative to this component's parent.
	* @param BoxMovability
	* The movability of the PhysicsBody.
	* @param CollisionLayers
	* The layers of the PhysicsBody.
	*/
	public void CreateBox(
		Transform RelativeTransform,
		Collision.PhysicsMotionType BoxMovability,
		Collision.Layer CollisionLayers)
	{
		CreateColliderOfType(BodyType.Box, RelativeTransform, BoxMovability, CollisionLayers);
	}

	/**
	* @brief
	* Creates a @ref Physics::SphereBody with the given parameters.
	*
	* @param RelativeTransform
	* The Transform of the PhysicsBody, relative to this component's parent.
	* @param SphereMovability
	* The movability of the PhysicsBody.
	* @param CollisionLayers
	* The layers of the PhysicsBody.
	*/
	public void CreateSphere(
		Transform RelativeTransform,
		Collision.PhysicsMotionType SphereMovability,
		Collision.Layer CollisionLayers)
	{
		CreateColliderOfType(BodyType.Sphere, RelativeTransform, SphereMovability, CollisionLayers);
	}

	/**
	* @brief
	* Creates a @ref Physics::CapsuleBody with the given parameters.
	*
	* @param RelativeTransform
	* The Transform of the PhysicsBody, relative to this component's parent.
	* @param CapsuleMovability
	* The movability of the PhysicsBody.
	* @param CollisionLayers
	* The layers of the PhysicsBody.
	*/
	public void CreateCapsule(
		Transform RelativeTransform,
		Collision.PhysicsMotionType CapsuleMovability,
		Collision.Layer CollisionLayers)
	{
		CreateColliderOfType(BodyType.Capsule, RelativeTransform, CapsuleMovability, CollisionLayers);
	}

	/**
	* @brief
	* Sets the activeness of the body
	* 
	* Active means the collider can interact with the physics system.
	* Inactive means it can't.
	*/
	public void SetActive(bool NewActive)
	{
		NativeFunction.CallNativeFunction("PhysicsComponentSetActive", typeof(SetBoolDelegate), [NativePtr, NewActive]);
	}

	/**
	* @brief
	* Casts this physics component shape from Start to End.
	* 
	* @param Start
	* The start Engine.Transform of the cast.
	* @param End
	* The end position of the cast.
	* @param Layers
	* The layers to check.
	* @param ObjectsToIgnore
	* These objects shouldn't be considered for the query.
	*/
	public Collision.HitResponse ShapeCast(Transform Start, Vector3 End, Collision.Layer Layers, WorldObject[] ObjectsToIgnore = null)
	{
		IntPtr[] ComponentPtrs;
		if (ObjectsToIgnore == null)
		{
			ComponentPtrs = [];
		}
		else
		{
			ComponentPtrs = new IntPtr[ObjectsToIgnore.Length];

			for (int i = 0; i < ObjectsToIgnore.Length; i++)
			{
				ComponentPtrs[i] = ObjectsToIgnore[i].NativePtr;
			}
		}

		return (Collision.HitResponse)NativeFunction.CallNativeFunction(
			"PhysicsComponentShapeCast",
			typeof(ShapeCastDelegate),
			[NativePtr, Start, End, Layers, ComponentPtrs, ComponentPtrs.Length]
			);
	}

	/**
	* @brief
	* Queries collision with this physics component's shape.
	* 
	* @param Where
	* The Engine.Transform of the query.
	* @param Layers
	* The layers to check.
	* @param ObjectsToIgnore
	* These objects shouldn't be considered for the query.
	*/
	public Collision.HitResponse PhysicsComponentCollisionCheck(Transform Where, Collision.Layer Layers, WorldObject[] ObjectsToIgnore = null)
	{
		IntPtr[] ComponentPtrs;
		if (ObjectsToIgnore == null)
		{
			ComponentPtrs = [];
		}
		else
		{
			ComponentPtrs = new IntPtr[ObjectsToIgnore.Length];

			for (int i = 0; i < ObjectsToIgnore.Length; i++)
			{
				ComponentPtrs[i] = ObjectsToIgnore[i].NativePtr;
			}
		}

		return (Collision.HitResponse)NativeFunction.CallNativeFunction(
			"PhysicsComponentCollisionCheck",
			typeof(CollisionCheckDelegate),
			[NativePtr, Where, Layers, ComponentPtrs, ComponentPtrs.Length]
			);
	}
}